pkg/fftypes/identity.go

```diff
@@ -16,9 +16,226 @@
+	FireFlyCustomDIDPrefix = "did:firefly:ns/"
+)
+
+type IdentityMessages struct {
+	Claim        *UUID `json:"claim"`
+	Verification *UUID `json:"verification"`
+	Update       *UUID `json:"update"`
+}
+
+// IdentityBase are the immutable fields of an identity that determine what the identity itself is
+type IdentityBase struct {
+	ID        *UUID        `json:"id"`
+	DID       string       `json:"did"`
+	Type      IdentityType `json:"type" ffenum:"identitytype"`
+	Parent    *UUID        `json:"parent,omitempty"`
+	Namespace string       `json:"namespace"`
+	Name      string       `json:"name,omitempty"`
+}
+
+// IdentityProfile are the field of a profile that can be updated over time
+type IdentityProfile struct {
+	Description string     `json:"description,omitempty"`
+	Profile     JSONObject `json:"profile,omitempty"`
+}
+
+// Identity is the persisted structure backing all identities, including orgs, nodes and custom identities
 type Identity struct {
+	IdentityBase
+	IdentityProfile
+	Messages IdentityMessages `json:"messages,omitempty"`
+	Created  *FFTime          `json:"created,omitempty"`
+	Updated  *FFTime          `json:"updated,omitempty"`
+}
+
+// IdentityCreateDTO is the input structure to submit to register an identity.
+// The blockchain key that will be used to establish the claim for the identity
+// needs to be provided.
+type IdentityCreateDTO struct {
+	Name   string       `json:"name"`
+	Type   IdentityType `json:"type,omitempty"`
+	Parent *UUID        `json:"parent,omitempty"`
+	Key    string       `json:"key,omitempty"`
+	IdentityProfile
+}
+
+// IdentityUpdateDTO is the input structure to submit to update an identityprofile.
+// The same key in the claim will be used for the update.
+type IdentityUpdateDTO struct {
+	IdentityProfile
+}
+
+// SignerRef is the nested structure representing the identity that signed a message.
+// It might comprise a resolvable by FireFly identity DID, a blockchain signing key, or both.
+type SignerRef struct {
 	Author string `json:"author,omitempty"`
 	Key    string `json:"key,omitempty"`
 }
+
+// IdentityClaim is the data payload used in a message to broadcast an intent to publish a new identity.
+// Most claims (except root orgs, where different requirements apply) require a separate IdentityValidation
+// from the parent identity to be published (on the same topic) before the identity is considered valid
+// and is stored as a confirmed identity.
+type IdentityClaim struct {
+	Identity *Identity `json:"identity"`
+}
+
+// IdentityVerification is the data payload used in message to broadcast a verification of a child identity.
+// Must refer to the UUID and Hash of the IdentityClaim message, and must contain the same base identity data.
+type IdentityVerification struct {
+	Claim    MessageRef   `json:"claim"`
+	Identity IdentityBase `json:"identity"`
+
+	// SignerRef lets us store back the message when broadcasting, but isn't part of the payload
+	SignerRef *Identity `json:"-"`
```

Comment:
Slightly confusing that this field is called "SignerRef" but is not of type "SignerRef"...

Reply:
Should it be IdentityRef as it is in the next struct below?

Reply:
It does get used, but it is confusing. I'll make the change to favor clarity in the structures, vs. consistency with the `Definition` interface

Reply:
It fell out in the wash in the end, as the claim processing logic actually sets the `messages.verification` ID on the `Identity` (not the verification processing logic). Thanks for spotting this cleanup.